Add the notion of equality to entities.

hsorby opened the pull request:
Looking to resolve #719 by defining equality between entities.

hsorby committed:
Start to add the notion of equality to entities.

hsorby commented:
I have started to address the problem to the find units bug.  People who are interested should comment on how it is shaping up.  Clearly this work is incomplete but discussion could be started before too much work is done.

nickerso commented on src/variable.cpp:
```diff
@@ -213,6 +215,19 @@ VariablePtr Variable::create(const std::string &name) noexcept
+        if (variable &&
+                this->initialValue() == variable->initialValue() &&
+                this->interfaceType() == variable->interfaceType()) {
+            return this->units() ? this->units()->equal(variable->units()) : true;
```
this would return `true` is this variable has not units set even if the other variable has units, right?

hsorby replied:
Yes, still more tests to be written to fully define what it means to be equal.

nickerso commented:
Not sure I completely grasp the inheritance happening here, but I think this looks like its going in the right direction and will be useful functionality.

kerimoyle commented:
I'm really confused by the idea here.  I can see that testing for equality means testing equality between what is represented (and not just where it's stored) ... but I still don't see that this will resolve the problems in #719 or #772 and #773.  Even if we implement an `equal` method everywhere (which I guess has a use case, even though it hasn't been shown yet), I still think that we absolutely have to have the option to locate (ie: add, remove, replace etc) by reference alone as well.  All I can see is a world of model debug nightmares and unexpected behaviour (not to mention unnecessary processing time!) if we use equality of content to identify items.  What am I missing ... ?

nickerso committed:
Merge branch 'develop' into equality

hsorby committed:
Merge in changes from 'prime/develop' fixing conflicts.

hsorby committed:
Progress towards adding equal function to all entities.

hsorby committed:
Adding equal logic to more entities.

hsorby committed:
Refuse to add the same entity repeatedly.

hsorby committed:
Revert changes to the private implementation of entities.

hsorby committed:
Merge branch 'equality' of github.com:hsorby/libcellml into equality

hsorby committed:
Add test for variables differing in units.

hsorby committed:
Add another test for variables with differeing units.

hsorby committed:
Add const where appropriate to find equal methods.

hsorby committed:
Cast unsigned long to signed long in vector offsetting.

hsorby committed:
Replace most int64_t with ssize_t (except sha1 calculation ones).

hsorby committed:
Address clang-tidy error.

hsorby committed:
Add a basic double comparison function.  Add note to add* methods on maximum number of additions assumed.

hsorby committed:
Consider more advanced nearlyEqual function.

hsorby committed:
Fix failing tests.

hsorby committed:
Correct clang tidy error.

hsorby committed:
Fix compilation errors.

hsorby committed:
Apply clang formatting to codebase.

hsorby committed:
Add Python test for equal.

hsorby committed:
Add doxygen documentation to doEqual and equal methods.

hsorby committed:
Remove whitespace from end of line in variable.h.

hsorby committed:
Fix up private doxygen documentation to be consistent.

hsorby committed:
Add test for equality of slightly modified HH model.

hsorby commented:
This is ready for review.

hsorby committed:
Minor comments added for add* methods.

nickerso commented on src/api/libcellml/entity.h:
```diff
@@ -109,12 +111,29 @@ class LIBCELLML_EXPORT Entity
+     * @brief Test to see if this entity is equal to the @p other entity.
+     *
+     * The attributes and properties of the entity are tested against the @p other
+     * entities attributes and properties and compared for equality.
```
entities -> entity's

hsorby replied:
22ed450

hsorby committed:
Update entity.h

kerimoyle commented on src/api/libcellml/component.h:
```diff
@@ -131,6 +131,9 @@ class LIBCELLML_EXPORT Component: public ComponentEntity, public ImportedEntity
      * was previously in a different component, it is moved to this one, and the
      * previous component's variable list is updated.
      *
+     * Note: We assume that less than 2,147,483,647 variables will be added.
```
fewer

hsorby replied:
d989210

kerimoyle commented on src/model.cpp:
```diff
@@ -49,20 +50,55 @@ struct Model::ModelImpl
+            }
+        }
+        if (unitsFound) {
+            // We are going to assume here that nobody is going to add more
```
Elsewhere we use the MAX_SIZE_T const to guard against this stuff?

hsorby replied:
No we don't use MAX_SIZE_T to guard against this stuff.

hsorby replied:
d989210

kerimoyle commented on src/reset.cpp:
```diff
@@ -197,4 +197,25 @@ ResetPtr Reset::clone() const
+{
+    if (Entity::doEqual(other)) {
+        auto reset = std::dynamic_pointer_cast<Reset>(other);
+        if (reset != nullptr && mPimpl->mOrder == reset->order() && mPimpl->mResetValue == reset->resetValue() && mPimpl->mResetValueId == reset->resetValueId() && mPimpl->mTestValue == reset->testValue() && mPimpl->mTestValueId == reset->testValueId()) {
```
This does a string comparison between MathML, which will fail if there are whitespace (or any other non-semantic) differences.  Esp if one set has been parsed, so come through our pretty filters compared to one entered through the API, this may not give expected behaviour!  It might be better to use the printer (or similar) to create standard forms of both strings before comparing them?

hsorby replied:
945d4ec

kerimoyle commented on src/reset.cpp:
```diff
@@ -197,4 +197,25 @@ ResetPtr Reset::clone() const
+        auto reset = std::dynamic_pointer_cast<Reset>(other);
+        if (reset != nullptr && mPimpl->mOrder == reset->order() && mPimpl->mResetValue == reset->resetValue() && mPimpl->mResetValueId == reset->resetValueId() && mPimpl->mTestValue == reset->testValue() && mPimpl->mTestValueId == reset->testValueId()) {
+            bool equal = true;
+            if (mPimpl->mTestVariable != nullptr && !mPimpl->mTestVariable->equal(reset->testVariable())) {
```
Would it make it run faster if you added (in the second position) a test for `reset->testVariable() != nullptr` ? No point in calling the function if it's null.  Also for the `variable` below.

hsorby replied:
I think the performance gains here would be so small that it would not be useful.  If you are looking to write performant efficient programs don't write them in C++.  The first test here is to only protect against segfaulting.

kerimoyle commented on src/units.cpp:
```diff
@@ -265,6 +268,103 @@ bool Units::isBaseUnit() const
+ *
+ *   https://bitbashing.io/comparing-floats.html
+ *
+ * @param a A double to test.
```
@c double ?

hsorby replied:
d989210

kerimoyle commented on src/variable.cpp:
```diff
@@ -213,6 +213,23 @@ VariablePtr Variable::create(const std::string &name) noexcept
 }
 
+bool Variable::doEqual(const EntityPtr &other) const
+{
```
I know it's covered in the very last function called (in the `Entity` class) but wouldn't it be more efficient to reject `nullptr` arguments when the `equal` functions are first called instead of going all the way down the call stack?

hsorby replied:
Yes it would be more efficient but C++ is not all about efficiency.

kerimoyle commented on src/variable.cpp:
```diff
@@ -213,6 +213,23 @@ VariablePtr Variable::create(const std::string &name) noexcept
+{
+    if (NamedEntity::doEqual(other)) {
+        auto variable = std::dynamic_pointer_cast<libcellml::Variable>(other);
+        if (variable != nullptr && mPimpl->mInitialValue == variable->initialValue() && mPimpl->mInterfaceType == variable->interfaceType()) {
```
Needs more brackets to be consistent with form elsewhere? Also below.
`((variable != nullptr )&& (mPimpl->mInitialValue == variable->initialValue()) && (mPimpl->mInterfaceType == variable->interfaceType())) `

hsorby replied:
Clang tidy approved this.

kerimoyle replied:
Haha, the stricter checks of @agarny may not ...

agarny replied:
Indeed, I would find it more readable to have:
```c++
if ((variable != nullptr)
    && (mPimpl->mInitialValue == variable->initialValue())
    && (mPimpl->mInterfaceType == variable->interfaceType())) {
```

kerimoyle commented on src/variable.cpp:
```diff
@@ -213,6 +213,23 @@ VariablePtr Variable::create(const std::string &name) noexcept
+        auto variable = std::dynamic_pointer_cast<libcellml::Variable>(other);
+        if (variable != nullptr && mPimpl->mInitialValue == variable->initialValue() && mPimpl->mInterfaceType == variable->interfaceType()) {
+            bool equal = true;
+            if (mPimpl->mUnits == nullptr && variable->units() != nullptr) {
```
This part confuses me a little after our discussions.  Why do you need to test that the meat of the `Units` is equal, if they can be uniquely assigned/defined by name alone?

hsorby replied:
I can't see how the fact that `Units` can be defined by name alone has anything to do with equality.  `Units` are equal if `equal` returns `true`.

kerimoyle commented on tests/bindings/python/test_component.py:
```diff
@@ -162,8 +162,8 @@ def test_remove_variable(self):
-        v1 = Variable()
-        v2 = Variable()
+        v1 = Variable("second")
+        v2 = Variable("meter")
```
This is changing what the test is doing.  I think it should be a new test, so that we can see what happens when you remove un-named variables.  Also, why name a variable with units? That's confusing ... ?

hsorby replied:
No, it is keeping the intent of the test the same.  Python tests are only for exercising the Python methods any functional testing should be done through the C++ tests.

kerimoyle commented on tests/bindings/python/test_model.py:
```diff
@@ -206,9 +206,9 @@ def test_replace_units(self):
-        self.assertFalse(m.replaceUnits(b, a))
-        self.assertFalse(m.replaceUnits(a, b))
-        self.assertFalse(m.replaceUnits(b, a))
+        self.assertTrue(m.replaceUnits(b, a))
```
I think this test needs to be made a lot clearer.  This is a weird case where `replaceUnits` takes arguments of old, new -> b doesn't exist in the model, so nothing should happen.  BUT since both are empty, the `equal` function will assume that the user means replace pseudo_b (ie: a) with itself ... so the references are switched, but the content is not?  It would demonstrate the behaviour more clearly if there was identical non-empty content in both units; as it is, this is just confusing and counter-intuitive.

hsorby replied:
I don't see how having identical non-empty content changes anything equal is still equal and the replacement would still return `True`.  Again the Python tests are for exercising the Python API functional examples should be made in the C++ tests.

kerimoyle commented on tests/component/component.cpp:
```diff
@@ -556,9 +556,10 @@ TEST(Component, addVariableMultipleTimes)
     // Adding a pip to the tomato.
```
Add not adding.

hsorby replied:
d989210

kerimoyle commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,995 @@
+
+TEST(Equality, invalidInput)
+{
+    libcellml::VariablePtr v1 = libcellml::Variable::create("variable");
```
auto?  also throughout

hsorby replied:
We don't have to use `auto` just because we can.  For me differences in the tests are a good thing.

kerimoyle commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,995 @@
+}
+
+TEST(Equality, variableNotEqualById)
+{
```
Need to test empty id vs non-empty id too?

hsorby replied:
I don't think so.  The test is for equal and not equal the string comparison which the id comparison is based on is well defined in this.

kerimoyle commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,995 @@
+    v1->setInitialValue(v3);
+    v2->setInitialValue(v4);
+
+    v1->setInterfaceType(libcellml::Variable::InterfaceType::PUBLIC);
```
Interfaces must be exact? Or would you argue that `PUBLIC` or `PRIVATE` ~= `PUBLIC_AND_PRIVATE` ?

hsorby replied:
Yes, interfaces must be exact.

kerimoyle commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,995 @@
+    u1->setSourceUnits(is, "ampere");
+    u2->setSourceUnits(is, "ampere");
+
+    u1->addUnit("meter");
```
See comment on python tests too - NB will not resolve to built-in units.

hsorby replied:
Sorry I don't understand this comment.

kerimoyle replied:
The spelling of `meter` is wrong.  It should be `metre` if you want it to represent built in units.  If not, I suggest you change the name to something more obviously custom?

kerimoyle commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,995 @@
+    libcellml::ImportSourcePtr is = libcellml::ImportSource::create();
+
+    u1->addUnit("second", 1.0, "second_id");
+    u1->addUnit("meter", -1.0, "meter_id");
```
these don't exist? Doesn't change the outcome of the test, but it may not be testing what you think it is?

hsorby replied:
For me it is testing what I want it to test.

kerimoyle replied:
`meter` becomes a user-defined base unit, NOT the built-in `metre` unit.  I think you should change the name to be more obvious it's not the built in unit, if that's what you intended, or correct the spelling so that it is?

kerimoyle commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,995 @@
+}
+
+TEST(Equality, unitsEqualPrecision3)
+{
```
If it fails with 2x, why do you need to test 3x as well?

hsorby replied:
82ed235

kerimoyle commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,995 @@
+
+    r1->setResetValue("reset value math");
+    r2->setResetValue("reset value math");
+
```
See comment in code too.  I think that in order to test these, we need to make sure that the whitespace isn't taken into account, especially since when setting it by parsing a model cleans the string that's stored, but when setting it through the API doesn't.

hsorby replied:
945d4ec

kerimoyle commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,995 @@
+    r1->setVariable(v1);
+    r2->setVariable(v2);
+
+    EXPECT_FALSE(r1->equal(r2));
```
Should all these tests also include the reverse test too for completeness?  ie: `r2->equal(r1)` ?

hsorby replied:
82ed235

kerimoyle commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,995 @@
+    cChild1->addComponent(cChild11);
+    cChild2->addComponent(cChild21);
+
+    EXPECT_TRUE(c1->equal(c2));
```
Making a comment so I remember to check:  is a child component equal to a non-child component?  ie: in this example, if you had another component that wasn't encapsulated, is it the same as one that is? ie: Does/should position in the hierarchy matter?
```
libcellml::ComponentPtr cFakeChild31 = libcellml::Component::create("child11");
EXPECT_WHAT(cFakeChild->equal(cChild11);
```

kerimoyle commented on tests/reset/reset.cpp:
```diff
@@ -274,6 +274,8 @@ TEST(Reset, removeResetInvalidArguments)
     libcellml::ComponentPtr c = libcellml::Component::create();
     libcellml::ResetPtr r1 = libcellml::Reset::create();
+    r1->setOrder(2);
+
```
I don't understand why this test needed to change?  If you wanted to show a different situation, why not add it?

hsorby replied:
This test had to change to show the same behaviour.  If `r1` is not different to `r2` then `r1` is removed later on when you ask to remove a `reset` that is equal to `r2`.

kerimoyle commented on tests/validator/validator.cpp:
```diff
@@ -195,6 +189,14 @@ TEST(Validator, unnamedAndDuplicateNamedVariablesWithAndWithoutValidUnits)
     v4->setUnits("dollars");
     v5->setName("mullah");
     v5->setUnits("$$");
+
```
Again I don't understand the need for the change?  Is there a difference in behaviour? (and if so, eek?) Also in other tests.

hsorby replied:
028240c

kerimoyle reviewed (changes requested):
- Some changes to tests needed (see comments)
- I'm not happy with testing string equivalence between MathML strings since we "tidy" up ones that come through the `Parser` but not through the API.  
- I'm also still not convinced about behaviour, but will write a test later to show that.

hsorby committed:
Revert changes to tests that shouldn't have been changed.

hsorby committed:
Merge branch 'equality' of github.com:hsorby/libcellml into equality

hsorby committed:
Address review comments. Add reverse equality test for completness.

hsorby committed:
Fixes to documentation and protection against adding very large numbers of entities.

hsorby committed:
Cast to compare unsinged integers and not a mix.

hsorby commented:
Unless I have missed anything the only thing outstanding is the MathML comparison.  For this comparison I can only see us doing a whitespace agnostic test, I cannot see us doing a semantic equivalent comparison. Does anyone think differently?

nickerso commented:
> Unless I have missed anything the only thing outstanding is the MathML comparison. For this comparison I can only see us doing a whitespace agnostic test, I cannot see us doing a semantic equivalent comparison. Does anyone think differently?

I think there are many issues in dealing with the MathML as strings so I would be very tempted to put this into the "it'll do for now" bucket and come back to this when we start dealing with the math in a more thorough manner. A whitespace agnostic comparison will only address one of the many potential sources for a string to be different, so just a straight out string comparison is fine by me at this point.

hsorby committed:
Rationalise the math equality into a function.

hsorby committed:
Rename internal math.{h,cpp} to mathml.{h,cpp}.

kerimoyle commented on src/mathml.h:
```diff
@@ -0,0 +1,35 @@
+/**
+ * @brief Compare math to determine if it is equal.
+ *
+ * Compare the given math strings to determin if they are equal or not.
```
Determine

hsorby replied:
a362d5e

hsorby committed:
Fix typo in compareMath doxygen documentation.

kerimoyle commented on src/componententity.cpp:
```diff
@@ -50,7 +50,7 @@ std::vector<ComponentPtr>::iterator ComponentEntity::ComponentEntityImpl::findCo
 {
     return std::find_if(mComponents.begin(), mComponents.end(),
-                        [=](const ComponentPtr &c) -> bool { return c == component; });
+                        [=](const ComponentPtr &c) -> bool { return c->equal(component); });
```
Should the function be called `equals` instead of `equal`?

hsorby replied:
c292dde

kerimoyle commented on src/mathml.cpp:
```diff
@@ -0,0 +1,26 @@
+namespace libcellml {
+
+bool compareMath(const std::string &math1, const std::string &math2)
+{
```
Are we really going to say that differences in whitespace in a mathml string are enough to render it non-equal?  That seems strange to me, especially as we "prettify" it internally?

hsorby replied:
945d4ec

kerimoyle commented on src/mathml.cpp:
```diff
@@ -0,0 +1,26 @@
+namespace libcellml {
+
+bool compareMath(const std::string &math1, const std::string &math2)
+{
```
Are we really going to say that whitespace differences in two mathml strings render them unequal?  I think this is a bit dangerous, especially as we 'prettify' the mathml strings internally?

hsorby replied:
Yes we are. 945d4ec

kerimoyle commented on src/units.cpp:
```diff
@@ -265,6 +268,103 @@ bool Units::isBaseUnit() const
+}
+
+/**
+ * @brief Decide if two doubles are nearly equal.
```
Should this have @c double and @c float?

hsorby replied:
I don't think so, it is not saying it is a double like the param documentation.

kerimoyle commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,1034 @@
+}
+
+TEST(Equality, variableNotEqualByUnitsOther)
+{
```
Now that you have both directions of testing equality in the test above, this one isn't required?

hsorby replied:
67ce1b8

kerimoyle commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,1034 @@
+    u1->addUnit("second", 1.0, "unit_id");
+    u1->addUnit("meter", 3.0);
+    u2->addUnit("meter", 3.0);
+    u2->addUnit("second", 1.0, "unit_id");
```
meter -> metre

hsorby replied:
8b58708

kerimoyle commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,1034 @@
+    libcellml::ParserPtr parser = libcellml::Parser::create();
+    libcellml::ModelPtr model = parser->parseModel(fileContents("generator/hodgkin_huxley_squid_axon_model_1952/model.cellml"));
+
+    EXPECT_EQ(size_t(0), parser->issueCount());
```
Somewhere here I'd expect a test that looks at equality between the different ways of creating a model:
- parsing
- through the API entirely
- cloning
- flattening
I suspect that this will show that we don't have good things happening in the MathML strings, since the Parser (I think?) does some reformatting, which is different to how the maths is stored if you simply add it to a component?  
In any case, for completeness I think these tests are needed.

hsorby replied:
945d4ec

nickerso replied:
just to confirm that flattening isn't a problem, perhaps add a test like

```
auto parser = libcellml::Parser::create();
auto model = parser->parseModel(fileContents("modelflattening/outsidecomponentequivalentvariables.xml"));
auto clonedModel = model->clone();
auto importer = libcellml::Importer::create();
importer->resolveImports(model, resourcePath("modelflattening/"));
model = importer->flattenModel(model);

importer->resolveImports(clonedModel,  resourcePath("modelflattening/"));
clonedModel = importer->flattenModel(clonedModel);

EXPECT_TRUE(model->equals(clonedModel));
EXPECT_TRUE(clonedModel->equals(model));
```

hsorby replied:
ef8426a

nickerso replied:
I'm good to resolve this one, but just want to check @kerimoyle is also good.

kerimoyle reviewed (dismissed):
A few unaddressed comments from previous reviews still to be seen?
Also:
- I don't think that MathML whitespace differences should render components or resets unequal?
- global search for `meter` vs `metre` spelling, so that the tests are clear when you're using built in units and when you're not. 
- It would be good to have tests that compare the different methods of creating the "same" model:
   - parsing
   - via the API entirely
   - cloning
   - flattening from imports

hsorby committed:
Merge branch 'develop' into equality

hsorby commented:
If reviewers could mark any comments that are resolved as resolved that would help finding comments that haven't been responded to.  Thanks.

hsorby committed:
Change 'meter' for 'metre'.

hsorby committed:
Rename 'equal' to 'equals'.

hsorby committed:
Remove duplicate test variableNotEqualByUnitsOther.

hsorby committed:
Start to add test for making sure parsing a model doesn't prettify the line endings.

hsorby committed:
Update python tests to use 'equals' method.

hsorby committed:
Add test to show math is treated as a straight string.

hsorby committed:
Remove MathML cleaning functions from parser.

hsorby committed:
Merge branch 'equality' of github.com:hsorby/libcellml into equality

hsorby commented:
(I think) all review comments have been addressed.  Ready for another go 'round.

nickerso commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,1060 @@
+    libcellml::UnitsPtr u1 = libcellml::Units::create("unitsA");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("unitsA");
+
+    libcellml::ImportSourcePtr is = libcellml::ImportSource::create();
```
not used so can remove.

hsorby replied:
7f152cf

nickerso commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,1060 @@
+    libcellml::UnitsPtr u1 = libcellml::Units::create("unitsA");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("unitsA");
+
+    libcellml::ImportSourcePtr is = libcellml::ImportSource::create();
```
not used.

hsorby replied:
7f152cf

nickerso reviewed (dismissed):
couple of minor comments, but looks good to me and I believe will improve our `find*` methods.

Do we need a test like `model->equals(model)` - would that show anything useful?

(oh, and would want to see testing pass before merging!)

agarny commented on tests/units/units.cpp:
```diff
@@ -2484,8 +2484,9 @@ TEST(Units, addUnitsMultipleTimes)
 
-    // Try to add the same units a second time.
+    // Try to add the same units a second time. Rejected.
     EXPECT_FALSE(model->addUnits(units));
```
As in `tests/variable/variable.cpp` (and above), why not add another test:
```c++
    EXPECT_EQ(size_t(1), model->unitsCount());
```

hsorby replied:
That test is there about two lines further down.

agarny replied:
Argh, was only looking at the changed code. Ok, easily resolved conversation! :)

agarny commented on src/api/libcellml/component.h:
```diff
@@ -131,6 +131,9 @@ class LIBCELLML_EXPORT Component: public ComponentEntity, public ImportedEntity
      *
+     * Note: We assume that fewer than 2,147,483,647 variables will be added.
+     * Undefined behaviour may occur if this assumption is violated.
+     *
```
I would remove this note (and all other similar ones). It should be obvious how many variables can be added by simply looking at `VariablePtr variable(size_t index) const`. From there, anyone a tiny bit familiar with programming should know that if that "assumption" is violated then undefined behaviour may occur.

Now, because we are using `size_t`, I would assume that fewer than 18,446,744,073,709,551,615 variables can be added since I am on a 64-bit system. So, if for some reasons or another (I have yet to go through the rest of this PR), we only want to support up to 2,147,483,647 variables then we shouldn't use `size_t`, but maybe `uint32_t`?

hsorby replied:
b657367

agarny commented on src/api/libcellml/component.h:
```diff
@@ -444,6 +450,8 @@ class LIBCELLML_EXPORT Component: public ComponentEntity, public ImportedEntity
 
     bool doIsResolved() const override; /**< Virtual method for implementing isResolved, @private. */
 
+    bool doEqual(const EntityPtr &other) const override; /**< Virtual implementation method for equals, @private. */
```
We should really not have this method (same for `doAddComponent ()`, `doSetImportSource()` and `doIsResolved()`) declared here. This looks really bad. We should only have our constructors and our PIMPL-related code. (Same elsewhere in our codebase.)

hsorby replied:
This may be the case but I don't think it is the role of this PR to address this.  If indeed it is seen as an issue then one should be created.  I don't see this PR being responsible for addressing this issue.

agarny replied:
Agreed, I was merely "reminding" ourselves of the issue (which we have discussed before). I have therefore updated issue #402 to account for this particular issue.

agarny commented on src/api/libcellml/entity.h:
```diff
@@ -23,6 +23,8 @@ limitations under the License.
 namespace libcellml {
 
+class EntityPrivate;
+
```
Not used/needed. Delete.

hsorby replied:
79741f3

agarny commented on src/api/libcellml/entity.h:
```diff
@@ -109,12 +111,29 @@ class LIBCELLML_EXPORT Entity
+     * @brief Test to see if this entity is equal to the @p other entity.
+     *
+     * The attributes and properties of the entity are tested against the @p other
+     * entity's attributes and properties and compared for equality.
```
Might want to add a comma before "and compared for equality".

hsorby replied:
79741f3

agarny commented on src/api/libcellml/entity.h:
```diff
@@ -109,12 +111,29 @@ class LIBCELLML_EXPORT Entity
+     * The attributes and properties of the entity are tested against the @p other
+     * entity's attributes and properties and compared for equality.
+     *
+     * If an entity has any children associated with it, then the order that children appear in
+     * is **not** taken into account.
```
Am I correct in understanding this as meaning that if an entity has children A and B, and another entity has children B and A, then they will be considered equal? That's certainly what I would expect.

hsorby replied:
Yes you are correct in your understanding and that is what the tests show.

agarny replied:
Yes, I saw that from the code and then the tests. I just couldn't be bothered going through my comments to find it again (I collapse files when I have reviewed them). :)

agarny commented on src/api/libcellml/entity.h:
```diff
@@ -109,12 +111,29 @@ class LIBCELLML_EXPORT Entity
+     * The parents of either entity (if they exist) are ignored when determining equality.
+     *
+     * @param other The entity to test against.
+     * @return @c true if this entity is equal to the other entity, @c false otherwise.
```
"... to the @p other entity..."?

hsorby replied:
79741f3

agarny commented on src/api/libcellml/importedentity.h:
```diff
@@ -24,6 +24,8 @@ limitations under the License.
 
 namespace libcellml {
 
+class ImportedEntityPrivate;
```
Not used/needed. Delete.

hsorby replied:
e4a20f0

agarny commented on src/component.cpp:
```diff
@@ -43,27 +45,82 @@ struct Component::ComponentImpl
+    for (const auto &variable : mVariables) {
+        bool variableFound = false;
+        size_t index = 0;
+        for (index = 0; index < unmatchedIndex.size() && !variableFound; ++index) {
```
I would put parentheses around `index < unmatchedIndex.size()`.

agarny commented on src/component.cpp:
```diff
@@ -43,27 +45,82 @@ struct Component::ComponentImpl
+                variableFound = true;
+            }
+        }
+        if (variableFound && index < size_t(std::numeric_limits<ptrdiff_t>::max())) {
```
Parentheses around `index < size_t(std::numeric_limits<ptrdiff_t>::max())`.

agarny commented on src/component.cpp:
```diff
@@ -43,27 +45,82 @@ struct Component::ComponentImpl
+    return true;
+}
+
+bool Component::ComponentImpl::equalResets(const ComponentPtr &other) const
```
Exactly the same method content as `equalVariables()` (and `Model::ModelImpl::equalUnits()`!). From what I can tell, it should be easy to refactor since both `mVariables` and `mResets` contain entities.

hsorby replied:
7615fd8

agarny commented on src/component.cpp:
```diff
@@ -473,4 +530,15 @@ bool Component::doIsResolved() const
+{
+    if (ComponentEntity::doEqual(other)) {
+        auto component = std::dynamic_pointer_cast<Component>(other);
+        if (component != nullptr && compareMath(mPimpl->mMath, component->math()) && mPimpl->equalResets(component) && mPimpl->equalVariables(component) && ImportedEntity::doEqual(component)) {
```
Readability. Replace with:
```c++
if ((component != nullptr) && compareMath(mPimpl->mMath, component->math())
    && mPimpl->equalResets(component) && mPimpl->equalVariables(component)
    && ImportedEntity::doEqual(component)) {
```

Better still, replace the `if` statement with:
```c++
return (component != nullptr) && compareMath(mPimpl->mMath, component->math())
       && mPimpl->equalResets(component) && mPimpl->equalVariables(component)
       && ImportedEntity::doEqual(component);
```

agarny commented on src/componententity.cpp:
```diff
@@ -273,4 +279,21 @@ std::string ComponentEntity::encapsulationId() const
+{
+    if (NamedEntity::doEqual(other)) {
+        auto componentEntity = std::dynamic_pointer_cast<ComponentEntity>(other);
+        if (componentEntity != nullptr && mPimpl->mEncapsulationId == componentEntity->encapsulationId() && mPimpl->mComponents.size() == componentEntity->componentCount()) {
```
Readability:
```c++
if ((componentEntity != nullptr)
    && (mPimpl->mEncapsulationId == componentEntity->encapsulationId())
    && (mPimpl->mComponents.size() == componentEntity->componentCount())) {
```

hsorby replied:
153a5ca

agarny commented on src/api/libcellml/entity.h:
```diff
@@ -109,12 +111,29 @@ class LIBCELLML_EXPORT Entity
 protected:
-    Entity(); /**< Constructor. */
+    Entity(); /**< Constructor, @private. */
+    virtual bool doEqual(const EntityPtr &other) const; /**< Virtual implementation method for equals, @private. */
```
This ought to be renamed `doEquals()` since it's related to `equals()`.

hsorby replied:
adf2c73

agarny commented on src/importedentity.cpp:
```diff
@@ -74,4 +76,14 @@ bool ImportedEntity::isResolved() const
+{
+    bool isImportLocal = isImport();
+    bool importMatches = isImportLocal == other->isImport();
+    if (isImportLocal && importMatches && mPimpl->mImportReference == other->importReference()) {
```
Readability:
```c++
if (isImportLocal && importMatches
    && (mPimpl->mImportReference == other->importReference())) {
```

hsorby replied:
153a5ca

agarny commented on src/mathml.h:
```diff
@@ -0,0 +1,35 @@
+ * @param math2 The second parameter to compare against parameter one.
+ * @return Return @c true if the @p math1 is equal to @p math2, @c false otherwise.
+ */
+bool compareMath(const std::string &math1, const std::string &math2);
```
The name `compareMath()` is misleading. I would expect it to behave in the same way as `std::string::compare()` (i.e. return a meaningful `int` value), which it does not. If we simply want to return a `bool` then I would rename this method to `equalMath()` or something like this, or maybe even `areEqual()` as we have done for `double` values in `utilities.h`?

This actually takes me to my second point: do we really need a `.cpp` and a `.h` file for such a trivial method? Couldn't we just add `areEqual(const std::string &str1, const std::string &str2)` to `utilities.h`? If we ever need to add more MathML-related methods later then we can have a dedicated `mathml.[cpp|h]` file, but for now there is really no need for one.

hsorby replied:
0dc2c0a

agarny commented on src/model.cpp:
```diff
@@ -49,20 +50,55 @@ struct Model::ModelImpl
+                        [=](const ImportSourcePtr &iS) -> bool { return iS->equals(importSource); });
+}
+
+bool Model::ModelImpl::equalUnits(const ModelPtr &other) const
```
See `Component::ComponentImpl::equalVariables()`.

hsorby replied:
7615fd8

agarny commented on src/component.cpp:
```diff
@@ -43,27 +45,82 @@ struct Component::ComponentImpl
+        size_t index = 0;
+        for (index = 0; index < unmatchedIndex.size() && !variableFound; ++index) {
+            size_t currentIndex = unmatchedIndex.at(index);
+            auto variableOther = other->variable(currentIndex);
```
Simplify:
```c++
auto variableOther = other->variable(unmatchedIndex.at(index));
```

hsorby replied:
7615fd8

agarny commented on src/model.cpp:
```diff
@@ -502,4 +541,15 @@ bool Model::fixVariableInterfaces()
+{
+    if (ComponentEntity::doEqual(other)) {
+        auto model = std::dynamic_pointer_cast<Model>(other);
+        if (model != nullptr && mPimpl->equalUnits(model)) {
```
Readability:
```c++
if ((model != nullptr) && mPimpl->equalUnits(model)) {
```

hsorby replied:
7615fd8

agarny commented on src/reset.cpp:
```diff
@@ -197,4 +199,25 @@ ResetPtr Reset::clone() const
+{
+    if (Entity::doEqual(other)) {
+        auto reset = std::dynamic_pointer_cast<Reset>(other);
+        if (reset != nullptr && mPimpl->mOrder == reset->order() && compareMath(mPimpl->mResetValue, reset->resetValue()) && mPimpl->mResetValueId == reset->resetValueId() && compareMath(mPimpl->mTestValue, reset->testValue()) && mPimpl->mTestValueId == reset->testValueId()) {
```
Readability:
```c++
if ((reset != nullptr) && (mPimpl->mOrder == reset->order())
    && compareMath(mPimpl->mResetValue, reset->resetValue())
    && (mPimpl->mResetValueId == reset->resetValueId())
    && compareMath(mPimpl->mTestValue, reset->testValue())
    && mPimpl->mTestValueId == reset->testValueId()) {
```

hsorby replied:
153a5ca

agarny commented on src/reset.cpp:
```diff
@@ -197,4 +199,25 @@ ResetPtr Reset::clone() const
+            } else if (mPimpl->mVariable == nullptr && reset->variable() != nullptr) {
+                equal = false;
+            }
+            return equal;
```
Readability:
```c++
if ((mPimpl->mTestVariable != nullptr)
    && !mPimpl->mTestVariable->equal(reset->testVariable())) {
    return false;
}

if ((mPimpl->mTestVariable == nullptr)
    && (reset->testVariable() != nullptr)) {
    return false;
}

if ((mPimpl->mVariable != nullptr)
    && !mPimpl->mVariable->equal(reset->variable())) {
    return false;
}

if ((mPimpl->mVariable == nullptr)
    && (reset->variable() != nullptr)) {
    return false;
}

return true;
```

or, if you dare:
```c++
return !(((mPimpl->mTestVariable != nullptr)
          && !mPimpl->mTestVariable->equal(reset->testVariable()))
         || ((mPimpl->mTestVariable == nullptr)
             && (reset->testVariable() != nullptr))
         || ((mPimpl->mVariable != nullptr)
             && !mPimpl->mVariable->equal(reset->variable()))
         || ((mPimpl->mVariable == nullptr)
             && (reset->variable() != nullptr)));
```

hsorby replied:
The second one is definitely not readable.

agarny replied:
Yes, I would definitely go with my first suggestion. My second suggestion was more for "fun". Hmm... 🤔

hsorby replied:
153a5ca

agarny commented on src/units.cpp:
```diff
@@ -22,7 +22,10 @@ limitations under the License.
 #include <algorithm>
 #include <cassert>
 #include <cmath>
+#include <cstdint>
+#include <cstring>
```
Are those needed? (Not on macOS at least.)

hsorby replied:
I think they are on Windows.

agarny replied:
Ok, probably are.

agarny commented on src/units.cpp:
```diff
@@ -265,6 +268,103 @@ bool Units::isBaseUnit() const
+bool nearlyEqual(const double a, const double b)
+{
+    const double fixedEpsilon = std::numeric_limits<double>::epsilon();
+    const ptrdiff_t ulpsEpsilon = 1;
```
Replace with:
```c++
static const double fixedEpsilon = std::numeric_limits<double>::epsilon();
static const ptrdiff_t ulpsEpsilon = 1;
```

hsorby replied:
2ded33f

agarny commented on src/units.cpp:
```diff
@@ -265,6 +268,103 @@ bool Units::isBaseUnit() const
+    const ptrdiff_t ulpsEpsilon = 1;
+
+    const double difference = fabs(a - b);
+    if (difference <= fixedEpsilon) {
```
Simplify:
```c++
if (fabs(a - b) <= fixedEpsilon) {
```

hsorby replied:
2ded33f

agarny commented on src/units.cpp:
```diff
@@ -265,6 +268,103 @@ bool Units::isBaseUnit() const
+ *
+ * @return @c true if the given doubles are considered close, @c false otherwise.
+ */
+bool nearlyEqual(const double a, const double b)
```
- Why not make this part of `Units::UnitsImpl`?
- Why specify `const`?

hsorby replied:
Because it does not require any member of `UnitsImpl`.

agarny replied:
Sure, but it's only used through UnitsImpl, so...?

Anyway, resolving this since this "issue" is moot now.

agarny commented on src/units.cpp:
```diff
@@ -265,6 +268,103 @@ bool Units::isBaseUnit() const
     return unitCount() == 0 && standardUnitCheck;
 }
 
+ptrdiff_t ulpsDistance(const double a, const double b)
```
- Why not make this part of `Units::UnitsImpl`?
- Why specify `const`?
- ULP? Unit of least precision?

hsorby replied:
Yes, ULP is units of least precision.
It shouldn't be part of `UnitsImpl` as it does not need to be.

agarny replied:
This "issue" is now moot, so resolving.

agarny commented on src/units.cpp:
```diff
@@ -265,6 +268,103 @@ bool Units::isBaseUnit() const
 
+ptrdiff_t ulpsDistance(const double a, const double b)
+{
+    const auto max = std::numeric_limits<ptrdiff_t>::max();
```
Replace:
```c++
static const auto max = std::numeric_limits<ptrdiff_t>::max();
```

hsorby replied:
2ded33f

agarny commented on src/units.cpp:
```diff
@@ -265,6 +268,103 @@ bool Units::isBaseUnit() const
+{
+    const auto max = std::numeric_limits<ptrdiff_t>::max();
+
+    // Max distance for NaN
```
Final full stop missing.

hsorby replied:
2ded33f

agarny commented on src/units.cpp:
```diff
@@ -265,6 +268,103 @@ bool Units::isBaseUnit() const
+    }
+
+    // If one's infinite and they're not equal, max distance.
+    if ((std::isinf(a) || std::isinf(b)) && (std::isinf(a) != std::isinf(b))) {
```
Replace with:
```c++
if (std::isinf(a) xor std::isinf(b)) {
```

hsorby replied:
2ded33f

agarny commented on src/units.cpp:
```diff
@@ -265,6 +268,103 @@ bool Units::isBaseUnit() const
+    ptrdiff_t ia;
+    ptrdiff_t ib;
+    memcpy(&ia, &a, sizeof(double));
+    memcpy(&ib, &b, sizeof(double));
```
Replace with:
```c++
static const int SIZE_OF_DOUBLE = sizeof(double);

ptrdiff_t ia;
ptrdiff_t ib;
memcpy(&ia, &a, SIZE_OF_DOUBLE);
memcpy(&ib, &b, SIZE_OF_DOUBLE);
```

hsorby replied:
2ded33f

agarny commented on src/units.cpp:
```diff
@@ -265,6 +268,103 @@ bool Units::isBaseUnit() const
+    // Return the absolute value of the distance in ULPs.
+    ptrdiff_t distance = ia - ib;
+    if (distance < 0) {
+        distance = -distance;
```
Replace with:
```c++
return -distance;
```

hsorby replied:
2ded33f

agarny commented on src/units.cpp:
```diff
@@ -265,6 +268,103 @@ bool Units::isBaseUnit() const
+ * @param b A @c double to test.
+ *
+ * @return @c true if the given doubles are considered close, @c false otherwise.
+ */
```
If we have some documentation for `nearlyEqual()` then why don't we have some for `ulpsDistance()`?

hsorby replied:
Because `nearlyEqual` is the entry function and `ulpsDistance` is only called by `nearlyEqual`.

agarny replied:
This issue is now moot, so resolving.

agarny commented on src/units.cpp:
```diff
@@ -265,6 +268,103 @@ bool Units::isBaseUnit() const
+    }
+
+    // If they are not the same sign then return false.
+    if ((a < 0) != (b < 0)) {
```
Replace with:
```c++
if ((a < 0.0) != (b < 0.0)) {
```

hsorby replied:
2ded33f

agarny commented on src/units.cpp:
```diff
@@ -265,6 +268,103 @@ bool Units::isBaseUnit() const
+        }
+    }
+    return false;
+}
```
I would rewrite this method as follows:
```c++
bool Units::doEqual(const EntityPtr &other) const
{
    if (!NamedEntity::doEqual(other)) {
        return false;
    }

    auto units = std::dynamic_pointer_cast<Units>(other);

    if ((units == nullptr)
        || (mPimpl->mUnits.size() != units->unitCount())
        || !ImportedEntity::doEqual(units)) {
        return false;
    }

    // Check unit definitions match.

    static const size_t PTRDIFF_T_MAX = size_t(std::numeric_limits<ptrdiff_t>::max());

    std::string reference;
    std::string prefix;
    double exponent;
    double multiplier;
    std::string id;
    std::vector<size_t> unmatchedUnitIndex(mPimpl->mUnits.size());

    std::iota(unmatchedUnitIndex.begin(), unmatchedUnitIndex.end(), 0);

    for (const auto &unitDefinition : mPimpl->mUnits) {
        bool unitFound = false;
        size_t index = 0;

        for (index = 0; (index < unmatchedUnitIndex.size()) && !unitFound; ++index) {
            size_t currentIndex = unmatchedUnitIndex.at(index);

            units->unitAttributes(currentIndex, reference, prefix, exponent, multiplier, id);

            if (nearlyEqual(unitDefinition.mExponent, exponent)
                && (unitDefinition.mId == id)
                && nearlyEqual(unitDefinition.mMultiplier, multiplier)
                && (unitDefinition.mPrefix == prefix)
                && (unitDefinition.mReference == reference)) {
                unitFound = true;
            }
        }

        if (!unitFound || (index >= PTRDIFF_T_MAX)) {
            return false;
        }

        unmatchedUnitIndex.erase(unmatchedUnitIndex.begin() + ptrdiff_t(index) - 1);
    }

    return true;
}
```

hsorby replied:
2ded33f

agarny commented on src/variable.cpp:
```diff
@@ -213,6 +213,23 @@ VariablePtr Variable::create(const std::string &name) noexcept
+            } else if (mPimpl->mUnits != nullptr) {
+                equal = mPimpl->mUnits->equals(variable->units());
+            }
+            return equal;
```
I would replace this with:
```c++
if (mPimpl->mUnits != nullptr) {
    return mPimpl->mUnits->equal(variable->units());
}

if (variable->units() != nullptr) {
    return false;
}

return true;
```

agarny commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,1060 @@
+    libcellml::UnitsPtr u1 = libcellml::Units::create("variable");
+
+    EXPECT_FALSE(v1->equals(u1));
+    EXPECT_FALSE(v1->equals(u1));
```
I reckon it should be:
```c++
EXPECT_FALSE(u1->equals(v1)); 
```
i.e. swap `u1` and `v1`.

hsorby replied:
e54b65a

agarny commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,1060 @@
+    libcellml::UnitsPtr u1 = libcellml::Units::create("units");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("units");
+
+    EXPECT_NE(u1, u2);
```
Does it need to be tested? (Same in `unitsNotEqualByNameOnly`.)

I wouldn't think so, but if so then why not for the variable-related tests above?

hsorby replied:
It is an exemplar, we don't actually cover every combination of possible circumstances in out tests.  Also we don't test everything in every test.  As long as it is tested somewhere and we can reasonably expect the behaviour to be the same , especially for the trivial cases.

agarny replied:
We are creating two objects, so they are bound to be different, but never mind.

agarny commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,1060 @@
+    EXPECT_NE(u1, u2);
+    EXPECT_TRUE(u1->equals(u2));
+    EXPECT_TRUE(u2->equals(u1));
+    EXPECT_TRUE(libcellml::Units::equivalent(u1, u2));
```
Does it need to be tested? (Same in `unitsNotEqualByNameOnly`.)

hsorby replied:
*Need* is quite subjective here. Here it shows expected behaviour.

agarny replied:
Ok, never mind.

agarny commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,1060 @@
+    libcellml::UnitsPtr u1 = libcellml::Units::create("unitsA");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("unitsA");
+
+    u1->addUnit("second", 1.0);
+    u2->addUnit("second", 1.0 + std::numeric_limits<double>::epsilon());
```
Just wondering: why do we want to allow these to be equal? I am not sure I like the idea that `unitsEqualPrecision3` and `unitsEqualPrecision4` pass.

nickerso replied:
this confirms that the way we are comparing real numbers is behaving as expected, right? which is a good thing to test...

agarny replied:
I guess I was a bit taken aback by the numbers `238487722399.29383` and `238487722399.29384` being considered nearly equal, even though they differ at the 5th decimal. However, I appreciate that those are big numbers, so a difference at the 5th decimal should indeed make those numbers nearly equal.

So, ok, I am fine with it. One thing, though, I think we should replace our existing `areEqual()` method in `utilities.cpp` with `nearlyEqual()` and, in fact, probably call that method `areNearlyEqual()`. In other words, I would remove `ulpsDistance()`, `nearlyEqual()` and `areEqual()`, and instead have:

```c++
bool areNearlyEqual(double a, double b)
{
    // Based on https://bitbashing.io/comparing-floats.html.

    // Classical equality check.

    static const auto Epsilon = std::numeric_limits<double>::epsilon();

    if (fabs(a - b) <= Epsilon) {
        return true;
    }

    // a and b must be of the same sign.

    if ((a < 0) != (b < 0)) {
        return false;
    }

    // The classical equality check wasn't successful, yet a and b are of the
    // same sign, so compute and check their ULPs distance.

    static const auto MaxUlpsDistance = std::numeric_limits<ptrdiff_t>::max();

    ptrdiff_t ulpsDistance = MaxUlpsDistance;

    if (!std::isnan(a) && !std::isnan(b) && !std::isinf(a) && !std::isinf(b)) {
        static const auto SizeOfDouble = sizeof(double);

        ptrdiff_t ia;
        ptrdiff_t ib;

        memcpy(&ia, &a, SizeOfDouble);
        memcpy(&ib, &b, SizeOfDouble);

        ulpsDistance = ia - ib;

        if (ulpsDistance < 0) {
            ulpsDistance = -ulpsDistance;
        }
    }

    static const ptrdiff_t UlpsEpsilon = 1;

    return ulpsDistance <= UlpsEpsilon;
}
```

Note that we don't check for `std::isinf(a) != std::isinf(b)` anymore. Indeed, it's not in the [original code](https://bitbashing.io/comparing-floats.html), so not sure how it ended up here, not least since there is no test for it.

hsorby replied:
Because nothing is exact when it comes to real number operations.  Here we are saying that we are setting our tolerance for arithmetic error to be 1 ULPs. `unitsEqualPrecision3` and `unitsEqualPrecision4` say the same thing.  I guess welcome to the world of number representation in computers.

agarny replied:
Yes, I am now fine with those tests, but what about https://github.com/cellml/libcellml/pull/785#discussion_r585481555?

hsorby replied:
`areEqual` and `areNearlyEqual` are different methods that serve different purposes.  So no they cannot be combined. `areNearlyEqual` is for numbers that (maybe) are the result of a mathematical operation.

hsorby replied:
I cannot see why you would not break that function down into its constituent parts.

hsorby replied:
There is a test for infinity, and there is a check for infinity in the original code.

agarny replied:
> `areEqual` and `areNearlyEqual` are different methods that serve different purposes. So no they cannot be combined. `areNearlyEqual` is for numbers that (maybe) are the result of a mathematical operation.

That's the thing, `areEqual()` and `areNearlyEqual()` actually serve the same purpose. `areEqual()` uses the classical equality check, which happens to be good enough for our current use of `areEqual()`. So, it's the same whether we use `areEqual()` or `areNearlyEqual()` for our current uses of `areEqual()`.

> I cannot see why you would not break that function down into its constituent parts.

Come on, it's a dead simple method. No need to break it down. It makes code maintenance unnecessarily "complicated".

agarny replied:
> There is a test for infinity, and there is a check for infinity in the original code.

I am clearly missing something since I can't see it...?!

hsorby replied:
In equality.cpp line 339.

hsorby replied:
> That's the thing, areEqual() and areNearlyEqual() actually serve the same purpose.

I disagree. `areEqual()` determines if in this case two doubles are exactly equal. `areNearlyEqual()` determines if two doubles are equal allowing for numerical error.

> Come on, it's a dead simple method. No need to break it down. It makes code maintenance unnecessarily "complicated".

I can't agree with that, calculating the ULPs distance is  a definite block of code that makes the function much more readable.  It in no way makes code maintenance unnecessarily complicated.

agarny replied:
> In equality.cpp line 339.

I guess you mean `TEST(Equality, unitsNotEqualInfinity)`? Anyway, I am talking about the original code at https://bitbashing.io/comparing-floats.html. I cannot see any `xor` operation there, just an `||` operation.



> > That's the thing, areEqual() and areNearlyEqual() actually serve the same purpose.
> I disagree. `areEqual()` determines if in this case two doubles are exactly equal. `areNearlyEqual()` determines if two doubles are equal allowing for numerical error.

`areEqual()` is currently defined as:
```c++
bool areEqual(double value1, double value2)
{
    return std::abs(value1 - value2) < std::numeric_limits<double>::epsilon();
}
```

which is the classical equality check in [my comment above](https://github.com/cellml/libcellml/pull/785#discussion_r585481555) (except for the fact that `areEqual()` uses `<` rather than `<=`).

> > Come on, it's a dead simple method. No need to break it down. It makes code maintenance unnecessarily "complicated".
> I can't agree with that, calculating the ULPs distance is a definite block of code that makes the function much more readable. It in no way makes code maintenance unnecessarily complicated.

It's a short function, called once, but never mind.

agarny commented on tests/equality/equality.cpp:
```diff
@@ -0,0 +1,1060 @@
+    EXPECT_FALSE(u2->equals(u1));
+}
+
+TEST(Equality, resetEqual)
```
Pretty much 100% of this test body is used in subsequent tests. It might be worth writing a method that initialises a reset-based test and then call whatever method needs calling to make the test fail. I.e.
```c++
TEST(Equality, resetEqual)
{
    libcellml::ResetPtr r1, r2;

    initialiseReset(r1);
    initialiseReset(r2);

    EXPECT_TRUE(r1->equals(r2));
    EXPECT_TRUE(r2->equals(r1));
}

TEST(Equality, resetNotEqualByOrder)
{
    libcellml::ResetPtr r1, r2;

    initialiseReset(r1);
    initialiseReset(r2);

    r2->setOrder(2);

    EXPECT_FALSE(r1->equals(r2));
    EXPECT_FALSE(r2->equals(r1));
}

...
```

hsorby replied:
e54b65a

agarny reviewed (commented):
Looks ok overall. Quite a few comments, but my main issue is something that was already in our codebase before and which is that there are some methods that have nothing to in our API, even though they are in the private section of a class. I am, for instance, talking about the `doEqual()` method (which probably ought to be renamed `doEquals()` in fact).

hsorby committed:
Add test for flattening.

hsorby committed:
Remove unused code from equality tests.

hsorby committed:
Remove note on limit to the number of entities can be added.

hsorby committed:
Address review comments for entity.h.

hsorby committed:
Remove class ImportedEntityPrivate;.

hsorby committed:
Address readability concerns.

hsorby committed:
Refactor equalUnits, equalVariables, equalResets.

hsorby committed:
Replace doEqual with doEquals.

hsorby committed:
Remove mathml.{h.cpp} and change compareMath to areEqual.

hsorby committed:
Tidy doEquals method in units.cpp.

hsorby committed:
Simplify equality tests for resets.

hsorby committed:
xor is not a cross-platform operator.  Replace with equivalent.

hsorby committed:
No else if after return in reset.cpp.

hsorby commented:
If reviewers could resolve conversations that have changes that meet their needs, that would help to see what is outstanding.  Many thanks.

hsorby committed:
Fix compilation errors on macOS.

hsorby committed:
Fix clang formatting errors.

kerimoyle commented on src/component.cpp:
```diff
@@ -177,8 +195,8 @@ bool Component::addVariable(const VariablePtr &variable)
+    // Prevent adding the same variable.
+    if (std::find(mPimpl->mVariables.begin(), mPimpl->mVariables.end(), variable) != mPimpl->mVariables.end()) {
         return false;
     }
```
I don't quite understand this?  Why would you even prevent the addition of duplicates any more, since two pointers to the same content are apparently interchangeable?

hsorby replied:
10a2411

kerimoyle commented on src/component.cpp:
```diff
@@ -298,8 +316,8 @@ bool Component::addReset(const ResetPtr &reset)
+    // Prevent adding the same reset.
+    if (std::find(mPimpl->mResets.begin(), mPimpl->mResets.end(), reset) != mPimpl->mResets.end()) {
         return false;
     }
```
Same comment as above - why are we preventing the addition of duplicate pointers, since every other operation now depends on their contents not their reference pointer?

hsorby replied:
10a2411

kerimoyle commented on src/importsource.cpp:
```diff
@@ -287,6 +287,17 @@ bool ImportSource::removeAllUnits()
+        }
+    }
+    return false;
+}
```
When testing for equal import sources, does it matter what they contain or only what they define?  ie
impA -> fileA
impB -> fileA
componentC, componentD both imported from impA
componentE imported from impB
Are impA and impB still equal?

hsorby replied:
Imports only describe where the information is coming from. So yes impA and impB are still equal.

kerimoyle commented on src/model.cpp:
```diff
@@ -120,8 +137,8 @@ bool Model::addUnits(const UnitsPtr &units)
-    if (hasUnits(units)) {
+    // Prevent adding the same units.
+    if (std::find(mPimpl->mUnits.begin(), mPimpl->mUnits.end(), units) != mPimpl->mUnits.end()) {
         return false;
```
Also here with the equality thing.  Why are we preventing it?

hsorby replied:
Because this is the behaviour we have already decided upon.

kerimoyle replied:
Well, no.  Everything in this PR is new behaviour, and the prevention of duplicates definitely belongs in the old behaviour set.  As far as I'm aware, we haven't ever discussed this in the new framework?  I find it very odd that you can delete, move, edit items based only on their content, but you can't add them that way?

hsorby replied:
The old behaviour is to stop you adding the same pointer to the list.  The change here maintains that behaviour.

kerimoyle replied:
I get it, but it doesn't make sense in the new context of thinking about items.  Why should the user have to care how we store the information?  Why can't you have two items the same on the list since their contents is now all we care about?

kerimoyle commented on src/model.cpp:
```diff
@@ -262,17 +279,20 @@ size_t Model::unitsCount() const
+    // Prevent adding the same import source.
+    if (std::find(mPimpl->mImports.begin(), mPimpl->mImports.end(), importSrc) != mPimpl->mImports.end()) {
         return false;
     }
```
Why prevent adding duplicates?

hsorby replied:
10a2411

kerimoyle commented on src/utilities.cpp:
```diff
@@ -1167,4 +1168,53 @@ std::string replace(std::string string, const std::string &from, const std::stri
 }
 
+bool equalEntities(const EntityPtr &owner, const std::vector<EntityPtr> &entities)
+{
```
Does this work when the order of entities is not matched?

agarny replied:
Yes, `unmatchedIndex` is used for this.

hsorby replied:
Yes

kerimoyle commented on src/utilities.h:
```diff
@@ -662,4 +675,18 @@ ConnectionMap createConnectionMap(const VariablePtr &variable1, const VariablePt
+ * @brief Test the given entities are equal to entities in owner.
+ *
+ * Test to see if all the entites given in @p entities are equal to
+ * entities in @p owner.  The order that the entities appear in is not
```
entities cf entites
@p owner in brief, @p entities in brief

hsorby replied:
6a63514

kerimoyle commented on src/utilities.h:
```diff
@@ -662,4 +675,18 @@ ConnectionMap createConnectionMap(const VariablePtr &variable1, const VariablePt
+ * @param entities The list of entities to equate.
+ *
+ * @return @c true if all the entities in @p entities are equal to entites in the @p owner.
+ */
```
Is the order important? Or just that the two sets are the same?

agarny replied:
The latter.

kerimoyle replied:
Then it should be spelled out in the docs.

hsorby replied:
The documentation of the method says that the order is not important.

agarny commented on src/reset.cpp:
```diff
@@ -199,22 +199,29 @@ ResetPtr Reset::clone() const
+            } else if ((mPimpl->mVariable == nullptr)
+                       && reset->variable() != nullptr) {
+                return false;
             }
```
I am pretty sure this will fail our CI tests. When there is a `return` statement within an `if` statement then we cannot have an `else` clause. IOW, it should be (also added some "missing" parentheses here and there):
```c++
if ((mPimpl->mTestVariable != nullptr)
    && !mPimpl->mTestVariable->equals(reset->testVariable())) {
    return false;
}

if ((mPimpl->mTestVariable == nullptr)
    && (reset->testVariable() != nullptr)) {
    return false;
}

if ((mPimpl->mVariable != nullptr)
    && !mPimpl->mVariable->equals(reset->variable())) {
    return false;
}

if ((mPimpl->mVariable == nullptr)
    && (reset->variable() != nullptr)) {
    return false;
}
```

hsorby replied:
This is not what is in the file, this is outdated.

agarny replied:
Yes, just came across your commit that does the above.

kerimoyle commented on tests/parser/file_parser.cpp:
```diff
@@ -89,35 +89,30 @@ TEST(Parser, parseModelWithComponentsWithMultipleMathElements)
+        "\t\t\t\t<cn xmlns:cellml=\"http://www.cellml.org/cellml/2.0#\" cellml:units=\"dimensionless\" type=\"e-notation\">0.5<sep/>1</cn>\n"
+        "\t\t\t</apply>\n"
+        "\t\t</math>\n";
 
```
What was the reason for the change in this test?  Do tabs equate to spaces now or no?

hsorby replied:
Because the test for math is now just a string comparison so the actual character is important, and no spaces are not equal to tabs.

kerimoyle replied:
So whitespace matters now ... wow, ok.  Resolving, but not particularly happy with this.

agarny commented on src/units.cpp:
```diff
@@ -278,7 +278,7 @@ ptrdiff_t ulpsDistance(double a, double b)
 
     // If one's infinite and they're not equal, max distance.
-    if (std::isinf(a) xor std::isinf(b)) {
+    if (!std::isinf(a) != !std::isinf(b)) {
```
`!X` is used to convert `X` to a `bool`, but `X` is already a `bool` here, so we can get away with just:
```c++
if (std::isinf(a) != std::isinf(b)) { 
```

This being said, the original code does a simple `||` operation not a `xor` one, so why a `xor` operation here?

kerimoyle replied:
.. because it only matters if they're different?  `||` doesn't give you that?

agarny replied:
Hmm. indeed, it only matters if they are different. (IOW, the original code could have been improved and that's what @hsorby did. Good good.)

hsorby replied:
The original code wasn't a simple `||`.

agarny replied:
?? I can't see a `xor` on `std::isinf()` at https://bitbashing.io/comparing-floats.html, or am I missing something? Anyway, I guess it doesn't matter.

hsorby committed:
Add annotation and correct typos in equalEntities documentation.

hsorby committed:
Rename 'nearlyEqual' to 'areNearlyEqual'.

hsorby committed:
Remove ! from in front of isinf.

hsorby commented:
Now that "we" have "remembered" that we can't add the same object twice I think all concerns have been addressed.

kerimoyle commented:
> Now that "we" have "remembered" that we can't add the same object twice I think all concerns have been addressed.

The parent thing @nickerso mentioned was never the issue.  We've already made the behaviour of "adding" something with an existing parent equivalent to removing it from that parent and adding it to the new one.  This is to do with using one set of rules for some operations (ie: identifying and manipulating using contents alone) vs another set for these adding operations (where the references are tested).

agarny commented:
> ... I think all concerns have been addressed.

Not as far as I am concerned, but well...

hsorby committed:
Remove the restriction about adding the same pointer twice.

hsorby committed:
Correct documentation for add* methods.

hsorby commented:
I have gone with @kerimoyle's thinking and that we shouldn't favour pointer location for some arbitrary reason.  The only thing that isn't done is import sources because I have already changed those in another (pending) PR.

agarny commented on tests/component/component.cpp:
```diff
@@ -119,7 +119,7 @@ TEST(Component, addAndCountChildren)
     child3->addComponent(child4);
     parent->addComponent(child3);
-    EXPECT_EQ(size_t(3), parent->componentCount());
+    EXPECT_EQ(size_t(4), parent->componentCount());
```
I would be curious to know what the resulting model is expected to be at this stage...

hsorby replied:
```
<?xml version="1.0" encoding="UTF-8"?>
<model xmlns="http://www.cellml.org/cellml/2.0#">
  <component name="parent"/>
  <component name="child1"/>
  <component name="child2"/>
  <component name="child3"/>
  <component name="child4"/>
  <component name="child3"/>
  <component name="child4"/>
  <encapsulation>
    <component_ref component="parent">
      <component_ref component="child1"/>
      <component_ref component="child2"/>
      <component_ref component="child3">
        <component_ref component="child4"/>
      </component_ref>
      <component_ref component="child3">
        <component_ref component="child4"/>
      </component_ref>
    </component_ref>
  </encapsulation>
</model>
```

For your curiosity.  I don't think the test needs this though.

agarny replied:
Agreed, we don't. This aside, to allow for this kind of behaviour doesn't make sense to me, but you guys seem to be ok with it, so whatever.

agarny commented on tests/component/component.cpp:
```diff
@@ -260,23 +263,23 @@ TEST(Component, removeComponentMethods)
-    // Still have component 'c2'.
-    EXPECT_EQ(size_t(1), c->componentCount());
+    // Can remove a second occurrence of "child1".
+    EXPECT_TRUE(c->removeComponent(c1));
```
Am I to understand that another call to `c->removeComponent(c1)` would work, but an additional one would fail? If so, I find it very confusing that after adding `c1` three times, I can remove it three times using the same reference. I could, at any point, modify `c1`, which means that when serialising the model, I will have three instances of `c1` in `c`...? Yup, I don't like it.

hsorby replied:
I don't like it either but I think @kerimoyle  is correct in that we are thinking here about the content that is contained.  It is odd behaviour but (and this for me is important) this is not something that a person would normally do.  It doesn't make sense to add the same entity multiple times.  It just doesn't make sense to add the same units multiple times to a model.

agarny replied:
If it doesn't make sense to add the same entity multiple times, then why do we allow it? What is fundamentally wrong with preventing `c1` from being added multiple times to `c`?

agarny commented:
> I have gone with @kerimoyle's thinking and that we shouldn't favour pointer location for some arbitrary reason. The only thing that isn't done is import sources because I have already changed those in another (pending) PR.

I am getting more and more confused by this PR, so leaving it to you guys to sort it out.

hsorby commented:
I think if this does go through in its current form then we should add an analysis for detecting repeated additions like we are no allowing.  This way at least people can analyse the model and find something we would consider to be a warning.

agarny commented:
> I think if this does go through in its current form then we should add an analysis for detecting repeated additions like we are no allowing. This way at least people can analyse the model and find something we would consider to be a warning.

So, you want to allow for duplicates and then you want to warn about duplicates because it duplicates don't make sense? Well, indeed, the whole thing doesn't make sense to me. I guess I need more sleep (I actually do).

kerimoyle commented:
> So, you want to allow for duplicates and then you want to warn about duplicates because it duplicates don't make sense?

The way I thought about it was solely in terms of content (since that's the point of this PR).  If a user wants to add many copies of a component or variable, and then adjust them later to make it valid, that's their call.  This region of adding things was the only place (that I could find ...) where we were effectively doing validation/prevention on an operation.  Nowhere else does that come into play ... and here it was only because of how libcellml was structured behind the scenes.  If this method doesn't make sense (and it might not!) then it should mean that more than just adding is changed.  Whatever philosophy we decide on should be applied consistently to all operations.

agarny commented:
> > So, you want to allow for duplicates and then you want to warn about duplicates because it duplicates don't make sense?
> 
> The way I thought about it was solely in terms of content (since that's the point of this PR). If a user wants to add many copies of a component or variable, and then adjust them later to make it valid, that's their call. This region of adding things was the only place (that I could find ...) where we were effectively doing validation/prevention on an operation. Nowhere else does that come into play ... and here it was only because of how libcellml was structured behind the scenes. If this method doesn't make sense (and it might not!) then it should mean that more than just adding is changed. Whatever philosophy we decide on should be applied consistently to all operations.

I agree that allowing/preventing duplicates has nothing to do with this PR, and that if we were to prevent them then we should ensure that it makes sense at the API-wide level. So, ok, happy to move on this one. (I have created issue #811 to prevent duplicates and look at the bigger picture.)

nickerso committed:
add missing header needed for windows compile

nickerso committed:
Add a couple of tests demonstrating multiple add behaviour.

With the change in behaviour introduced with the ability to add entities multiple times
it would seem useful to test for expected behaviour in case users come to depend on this
funky use case continuing to function consistently.

hsorby committed:
Merge pull request #12 from nickerso/equality

Windows compile and new tests

nickerso reviewed (dismissed):
While I'm not convinced that the way this PR changes the multiple adding of entities, I do think that can be left to #811 to solve. In 77a7a77 I did add a couple of tests which show the consequences of this just to make sure that we don't unknowingly change behaviour that users might come to rely on.

hsorby committed:
Apply clang format to src/variable.cpp

agarny committed:
Utilities: reworked areEqual() and replaced some areEqual() calls with areNearlyEqual() calls.

agarny committed:
Utilities: reworked areEqual() and replaced some areEqual() calls with areNearlyEqual() calls.

Make Linux happy.
